### ISSUE TITLE
Reactive APIs use case

### ISSUE THREAD

**Issue by sdeleuze** [Kotlin/kotlin-coroutines#43]
While I see how the proposed approached will work and bring value with future types like `CompletableFuture` or with synchronous stream transformation API like `java.util.stream.Stream`, I am wondering how it could work with Reactive types, like the one provided by [Reactor](http://projectreactor.io/), [RxJava](https://github.com/ReactiveX/RxJava) / [RxKotlin](https://github.com/ReactiveX/RxKotlin) and more generally with any API based on [Reactive Streams](http://www.reactive-streams.org/).

I have tried to describe the differences between `CompletableFuture` / `java.util.stream.Stream` and Reactive types like Reactor `Flux` / `Mono` types or RxJava `Observable` / `Single` types in this [Understanding Reactive types](https://spring.io/blog/2016/04/19/understanding-reactive-types) blog post since there is a lot of confusion on this topic.

Some of my concerns:
 - How do you handle asynchronous non-blocking streams operations like `flatmap()`?
 - How do you deal with Reactive Streams Backpressure?
 - Are Kotlin coroutines relevant for dealing with Reactive Streams error signal (where the exception is not thrown by current method but emitted as en error signal)?
 - How does that works when you want to isolate blocking API in specific thread pools with RxJava `subscribeOn()` / `observeOn()` operators?
 - Is the coroutine version really more easy to write/read than fluent API style transformation à la RxJava/Reactor for pipeline with significant number of transformation?

For the use cases I have in mind, it does not only concern integrating some Reactive APIs in a regular blocking architecture, but also about supporting native Reactive frameworks like the upcoming Spring Framework 5.

To illustrate what I mean, see:
 - These [A lite Rx API for the JVM](https://speakerdeck.com/sdeleuze/a-lite-rx-api-for-the-jvm) slides
 - [spring-reactive](https://github.com/spring-projects/spring-reactive)
 - [spring-reactive-playground](https://github.com/sdeleuze/spring-reactive-playground)
 - [Cloud Foundry Java client API](https://github.com/cloudfoundry/cf-java-client#cloudfoundryoperations-apis) based on Reactor 2.5
 - MongoDB [Reactive Streams](http://mongodb.github.io/mongo-java-driver-reactivestreams/) and [RxJava](http://mongodb.github.io/mongo-java-driver-rx/) drivers
 - [Couchbase RxJava driver](http://docs.couchbase.com/developer/java-2.0/observables.html)

**Comment by orangy** [comment]
It's in no way complete solution, and I even didn't prototype it with something real but my personal thinking was along the lines of this:
```
fun Observable<Observable<String>>.flatMap() = observable { // coroutine, creates and returns Observable
   val item = await(this) // subscribes, calls continuation on each 
   val itemItem = await(item) // subscribes, call continuation on each subitem
   push(itemItem)
}
```
So, `observable` is a function receiving corouting, `await` is suspension function, which somehow remembers state machine state and can call continuation several times. This is yet TBD, and it's a little bit confusing that part of your code can be run several times, unexpectedly. `push` is normal function, sending value to the observable.

**Comment by sdeleuze** [comment]
I understand what you mean, but `Observable` is IMO much more than an async dual to `Iterable`. It allows to deal with stream of data with synchronous (`map()`) or asynchronous (`flatMap()`) transformations, provide a wide range of operators that modify how you handle a stream of data (`Observable`), a single data (`Single`) or just a complete signal (`Completable`), allow you to control how much data the subscriber want to request, allow you to manage how you integrate blocking API and threading with `subscribeOn()` and `observeOn()` and a lot of other operations that allow you to define a pipeline of operation and control very precisely that async behavior.

`yeld`, `await`, and continuations is another abstraction, that can make sense we you just deal with `CompletableFuture` or `Stream`, but there are [good reasons](https://spring.io/blog/2016/04/19/understanding-reactive-types) why JVM ecosystem is moving to Reactive architectures and not just using these types. And my gut feeling is that while we could find some use cases where these approaches can be complementary (`await` on `Single` type for example), they are 2 different abstractions for async processing, mostly exclusive.

To quote [this C# await + Rx blog post](http://log.paulbetts.org/rx-and-await-some-notes/):

> **Crap, when do I use async/await and when do I use Rx?**
> Use whatever makes more straightforward code, but here's some suggestions:
>  - Wire everything up and do high-level stuff with Rx
>  - async/await is sometimes better when you have early returns in methods
>  - Do UI stuff in Rx, don't await events

As one of the developer of the Reactive support of Spring Framework 5, what I do like in Kotlin is that it is not opinionated too much about this topic, and is a perfect fit with Reactive API since Kotlin extensions provide the perfect way to extend fluent API à la RxJava (something that is not possible at all in Java), provide a very good functional programming support, etc.

My concern is that `yeld`, `await`, and continuations may be a "false good idea". It is an attempt that try to keep developers working in imperative programming world while they are not. I think it works for simple use cases, but won't be a good idea when you want to let the control to the developer about how he manages the demand (backpressure), defining a pipeline of transformation and deal with event based API. A Reactive API should be Reactive end to end, else it does not work.

I don't say there is nothing to solve, Reactive APIs are for example difficult to debug, and we have currently no way to avoid exposing 100+ operators to the user on RxJava `Observable` or Reactor `Flux`. But Reactive APIs is currently the only way to tackle globally the async and stream handling, it works in a wide range of programming languages, and Reactive Streams will be part of Java 9.

I think I prefer the current status where Kotlin + Rx are perfect used together rather than adding new magic capabilities, that will hide too much how async processing works.

**Comment by orangy** [comment]
> My concern is that yeld, await, and continuations may be a "false good idea".
… for Rx. From what I see from your comment, you can just skip coroutines for Rx and not use them at all.

**Comment by sdeleuze** [comment]
Indeed, I could simply not use them, but the idea behind my comment is that I have some concerns about the confusion that Kotlin coroutines could add for users if they are just syntaxic sugar for `Future` and `Stream` without being really useful for Reactive APIs like in C# world.

Maybe that could be nice to see if Kotlin can do better than C# by making Kotlin coroutines play well with Reactive APIs. And I think there is at least 3 main topics here:

1) Make sure that Kotlin coroutines are suitable for all behavior of async types:
 - Push-based: `CompletableFuture` or Reactive Streams with `request(Long.MAX_VALUE)`
 - Pull-based: `Stream` or Reactive Streams with `request(1)`
 - Dynamic push-pull strategy: regular Reactive Streams backpressure
Since Reactive Streams is becoming the defacto standard for Reactive semantics, maybe there is something to do on supporting natively Reactive Streams Publisher (both reactive-streams.org or JDK9 flavors)

2) How `await` could play well with RxJava `Single` / `Completable` and Reactor `Mono` types?

3) How can Kotlin coroutines could be attractive for non trivial RxJava `Observable` or Reactor `Flux` transformation pipelines?

**Comment by orangy** [comment]
Do you have any ideas about how it could look like? Is there anything to "solve" in Rx world, like async/await coroutines solves CPS lambda nesting hell for traditional async?

**Comment by cypressious** [comment]
Since C# has both, Coroutines and Reactive Extensions, how does it deal with it?

From a quick Google search I could find the following articles:

http://rehansaeed.com/reactive-extensions-part5-awaiting-observables/: You can `await` one item of an `Observable`, be it the first, last or single.

http://log.paulbetts.org/rx-and-await-some-notes/: Async SelectMany (equivalent to `flatMap`): Instead of creating one `Observable` per item, you create a task per item. If I understand correctly, this is equivalent to `flatMap { Observable.fromCallable { /* download string */ } }`. The benefit compared to using a simple `map { /* download string */ }` is that `flatMap` can run in parallel.

Actually, this isn't specific to `flatMap`. You could create an async factory method `Observable.fromAsync` that asynchronously `yield`s items into the stream. Currently, you would have to use `Observable.create` for this. Or you could do the same with `Observable.fromIterable` where the iterable is created using an async generator as demonstrated in http://stackoverflow.com/questions/10116573/preferred-method-for-generating-an-iobservablestring-from-a-stream.

I'm sure there are more examples but I just wanted to demonstrate that combining coroutines with Rx can be useful.

**Comment by Alxandr** [comment]
C# allows you to await anything (by providing an (extension-)method named `GetAwaiter` (iirc), the point being that it's convention based. Rather, I think a much more interesting case to look at is how F# deals with seq/observable/async. For instance, the following is an async in F#

```fsharp
let myAsync = async {
  let! foo = bar
  return! doSomeAsync(foo)
}
```

This compiles to the following (ish):

```fsharp
let myAsync = async.Bind(bar, fun foo ->
  async.ReturnFrom(doSomeAsync(foo)))
```

In F# you have a few rules of what expressions transform into what method calls on a "builder", and then you can create your own builders as you want to.

Some examples:
```fsharp
let mySeq = seq {
  yield 1
  yield 2
}

let myAsyncSeq = asyncSeq {
  let! tweets = fetchTweets() // async, break point
  for tweet in tweets do
    yield tweet
}
```

**Comment by elizarov** [comment]
With a revised coroutines in 1.1-M04 one can define an extension to iterate on observables in a straightforward way, so that the following code just works: 

```
fun Observable<Observable<String>>.flatMap(): Observable<String> = asyncRxGenerate {
    for (a in this@flatMap)
        for (b in a)
            emit(b)
}
```

Moreover, all the back-presse is automatically handled and you don't need to write any hairy and error-prone code to handle it.

The implementation of `asyncRxGenerate` coroutine builder with `emit` support can be found here: 
https://github.com/elizarov/kotlinx.coroutines/blob/asyncRxGenerate/kotlinx-coroutines-rx/src/main/kotlin/asyncRxGenerator.kt
It is pretty worked out and has full support of Rx back-pressure protocol. It generates as many items and consumer requests.

Implementation for `for (x in observable)` is a separate extension and can be found here: https://github.com/elizarov/kotlinx.coroutines/blob/master/kotlinx-coroutines-rx/src/main/kotlin/asyncRxIterator.kt
This one requests items one by one and actually has a one-item request-ahead buffer inside (despite the fact that its variable is called `rendezvous`).

**Comment by ilya-g** [comment]
> the following code just works

It may just work, but it isn't the [`flatMap`](http://reactivex.io/documentation/operators/flatmap.html) operation. 

Since the execution inside of coroutine is serial, elements of the second `a` could not be emitted before elements of the first `a` complete, and thus it behaves more like `map` + [`concat`](http://reactivex.io/documentation/operators/concat.html), while `flatMap` is `map` + [`merge`](http://reactivex.io/documentation/operators/merge.html).

**Comment by elizarov** [comment]
Good catch. The current version of `asyncRxGenerate` does not support `emit` from multiple coroutines, but it looks that it can be improved without a significant penalty, so that emits from multiple coroutines are properly supported. With this improvement, the actual `flatMap` will be represented as:

```
fun Observable<Observable<String>>.flatMap(): Observable<String> = asyncRxGenerate {
    for (a in this@flatMap)
        async { // just start a concurrent coroutine for each item with any simple builder
            for (b in a)
                emit(b)
        }
}
```
